Add enhancement for collecting installer telemetry metrics.

patrickdillon opened the pull request:
Authored together with @rna-afk 

This is a first draft of a design for collecting telemeter metrics in the installer.

patrickdillon commented:
@abhinavdahiya in addition to other feedback, please advise on reviewers/approvers

s-urbaniak commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,280 @@
+```
+# HELP cluster_installation_duration keeps track of the count of the duration of individual stages of the command execution in buckets with the result of the command.
+# TYPE cluster_installation_duration counter
+cluster_installation_duration{provisioning="5", infrastructure="5", bootstrap="5", api="5", operators="5" success="true"} 1
````
I suppose this means duration for `provisioning`, `infrastructure`, etc. As lili mentioned those label values should not be arbitrary (i.e. the actual duration), but rather be represented as separate metrics.

patrickdillon replied:
@s-urbaniak yes that is the intended meaning of this metric. The idea is to use label values as discussed here: https://github.com/openshift/enhancements/pull/205#discussion_r377123168

If we broke these into separate samples we would lose the correlation between the different stages, but perhaps that is ok. I think it is important that we know whether the samples come from a successful or failed installation, so the samples would look something like this:
```
cluster_installation_duration{provisioning="5", success="true"} 1
cluster_installation_duration{bootstrap="5", success="true"} 1
```
We would still have to use the label builder approach to delay creating metrics until we know the result of the command.

s-urbaniak commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,280 @@
+```
+# HELP cluster_installation_modification lists all the assets that were changed from the original value by the user before the create command was called.
+# TYPE cluster_installation_modification counter
+cluster_installation_modification{asset="cluster-config.yaml,cluster-scheduler-02-config.yml"} 1
````
Agreed, the upper bound of label values must be known in forehand.

patrickdillon replied:
I have updated the enhancement to provide more context. I think we have potentially ~30 values, with a reasonable expectation it could be much lower. Let me know if that sounds like a reasonable bound.

s-urbaniak commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,280 @@
+
+```
+// AddToMetric adds a sample to a metrics vector.
+AddToMetric(metric string, labels map[string]string, value string)
````
Just a general comment. As pointed out above there are a couple of best practice violations expressed above which would make it very easy to be used using this wrapping API. What speaks against using the Prometheus API directly?

patrickdillon replied:
This function `AddToMetric` could be replaced by a direct call to the Prometheus API. The reason for creating a wrapper is: to isolate Prometheus dependency in the installer to the new `metrics` package and create a uniform API within the installer base. 

Does the need for `AddLabelValue` make sense? 

> here are a couple of best practice violations expressed above 

If I haven't addressed any of those please let me know and we can discuss further.

patrickdillon commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,280 @@
+```
+# HELP cluster_installation_invocation represents a single command run by the OpenShift installer and related outcomes.
+# TYPE cluster_installation_invocation counter
+cluster_installation_invocation{command="create", target="cluster",os="linux",platform="AWS",result="ClusterComplete",return="0",version="4.2",duration="25"} 1
````
The idea is to bound cardinality by binning the values on the installer client side, similar to a histrogram but not inclusive. e.g. we could have values that represent that installation took: less than 15 minutes, 20 minutes, 25 minutes, 30 minutes, 35 minutes, 40 minutes, more than 45 minutes.

So that example would have 7 potential values. We would discuss what values would be best.

If that doesn't resolve the cardinality issues let me know and we can discuss alternative approaches.

LiliC replied:
Just having a bound cardinality sounds good to me, but I think that depends on how much you can send via telemetry, there are certain budgets per component. 


It's also depends on what is actually useful to you when analyzing this data. If 5 minute interval values are important then this makes sense. The same goes for all the rest of the labels.

bwplotka replied:
Why not create exactly a histogram format? This looks like something custom: `duration` vs official `le` label.

patrickdillon replied:
> Why not create exactly a histogram format? This looks like something custom: duration vs official le label.
+1 see: https://github.com/openshift/enhancements/pull/205#discussion_r381388713

patrickdillon commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,280 @@
+# HELP cluster_installation_invocation represents a single command run by the OpenShift installer and related outcomes.
+# TYPE cluster_installation_invocation counter
+cluster_installation_invocation{command="create", target="cluster",os="linux",platform="AWS",result="ClusterComplete",return="0",version="4.2",duration="25"} 1
+cluster_installation_invocation{command="create", target="cluster",os="linux",platform="AWS",result="ClusterComplete",return="0",version="4.2",duration="22.5"} 1
```
Correct. Updated the enhancement.

patrickdillon commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,280 @@
+cluster_installation_invocation{command="create", target="cluster",os="linux",platform="AWS",result="APIFailure",return="1",version="4.2",duration="30"} 1
+cluster_installation_invocation{command="destroy", target="cluster",os="linux",platform="AWS",result="Success",return="1",version="4.2",duration="10"} 1
+```
+The `result` label denotes a command outcome:
````
No. I updated the enhancement to clarify. We are still working on determining more values. Are there cardinality concerns? If so, what would be an acceptable number of values?

lilic replied:
They are concerns more on the telemetry server side of things, I think it would be good to know the final number of series you would be sending.  cc @brancz @smarterclayton

patrickdillon commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,280 @@
+Either of these proposed metrics would be pushed only upon successful completion of the `create cluster` command.
+
+#### Metric Labels
+Collecting labels into a single metric sample and pushing once means that each sample represents a single installer invocation. When a sample is pushed to the aggregation pushgateway metrics with matching labels will be incremented:
```
One push per invocation which would contain somewhere between 1-10 samples depending on the final design.

patrickdillon commented:
@LiliC @s-urbaniak thank you so much for taking a look. I have tried to address your comments.

abhinavdahiya commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+
+### Non-Goals
+
+1. The metrics collected will be aggregate metrics so metrics will not be relevant to individual installer invocations and will not display detailed data such as log or error messages
```
this is a little confusing.

bwplotka replied:
I agree, it's confusing for me as well. Does it mean that it's non goal to make it aggregated (reduce cardinality)? I guess the opposite?

abhinavdahiya commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,300 @@
+cluster_installation_invocation{command="create", target="cluster",os="linux",platform="AWS",result="ClusterComplete",return="0",version="4.2",duration="22.5"} 1
+cluster_installation_invocation{command="create", target="cluster",os="linux",platform="AWS",result="APIFailure",return="1",version="4.2",duration="30"} 1
+cluster_installation_invocation{command="destroy", target="cluster",os="linux",platform="AWS",result="Success",return="1",version="4.2",duration="10"} 1
+```
````
can we have the value of the metric be the return code? return codes are ints so we should be able to use that?

also all the values are 1? what are other valid values?

rna-afk replied:
Right now, the values are fixed with 1 for success and 0 for failure. The idea to use this return code label was to have specific error codes for specific errors that the installer faces during it's invocation. We might have only 0 or 1 now but in the future, we could designate different error codes that denote the exact error that the installer faced.

patrickdillon replied:
> Right now, the values are fixed with 1 for success and 0 for failure. 

This refers to the values for the return code. In response to: 

> also all the values are 1? what are other valid values?

The sample values are always 1, because the sample represents one invocation of an installer command. The aggregation push gateway will take the `1` value and add it to the value that is currently in Prometheus, so Prometheus will have a running count.

simonpasquier replied:
Is the `return` label really needed? If `result="Success"` then `return` is always 1 else it is 0.

abhinavdahiya commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+One sample would be pushed per command invocation.
+
+#### cluster_installation_modification
+The `cluster_installation_modification` metric represents whether a user modified an installer created asset prior to running the `create` command.
```
I think instead of having per file modification,

1) we track if modification was made
2) we define categories, like machines, machineconfigs, scheduler, network, ingress, dns so that we are tracking use-cases and not noise.

abhinavdahiya commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,300 @@
+```
+If the above example represents the state of our Prometheus database, it would indicate that there have been 100 successful cluster installations on AWS that took 25 or fewer minutes.
+
+The values for these labels may only become available at different times during an installation or command execution. Therefore the installer instrumentation must handle building labels by collecting key value pairs as they become available.
````
hmm, whose going to run or do this aggregation?

patrickdillon replied:
The idea is to run the [Weaveworks Aggregation Pushgateway](https://github.com/weaveworks/prom-aggregation-gateway) I am not sure which team would administer the pushgateway. This is discussed a little in Open Questions.

abhinavdahiya commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+If the above example represents the state of our Prometheus database, it would indicate that there have been 100 successful cluster installations on AWS that took 25 or fewer minutes.
+
+The values for these labels may only become available at different times during an installation or command execution. Therefore the installer instrumentation must handle building labels by collecting key value pairs as they become available.
+
```
hmm, we have 3 metrics being pushed from single invocation of the installer. it is possible to tag all these pushed together so that we can at least see all of them together?

most of the aggregation logic can drop/filter this label to reduce cardinality, but maybe just storing these would help in case of directed _pointed_ debugging?

patrickdillon replied:
As far as I can tell, the time series nature of Prometheus would make this approach problematic. As soon as we add a unique identifier in either a label **or** value, that metric would become its own time series, with a metric value that would never change. You could query over the metrics to aggregate, but you would not be able to reduce the cardinality. 

We should try to come up with some examples to illustrate/verify this conclusion.

abhinavdahiya commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+#### Internal Installer Design
+A new metrics package would be created to handle the collection of metrics, building labels, and pushing to the aggregation pushgateway. 
+
+As required label values are gathered throughout many installer packages,  the `metrics` package would maintain state package-scoped variables (feedback please: package-scoped variables may be generally frowned upon but is this a valid use case?).
```
i think it's important that the package allow initiating a sink for specific purpose.

- but we can start with all metrics being collected into the global sink.
- an alternative would be initiate a sink per new asset store to reduce use of global state.

abhinavdahiya commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,300 @@
+// AddToMetric adds a sample to a metrics vector.
+AddToMetric(metric string, labels map[string]string, value string)
+```
+
````
internal implementation should also probably also include

- how to stop the metrics
- exactly when the metrics are pushed.

rna-afk replied:
```suggestion
// StopMetrics is used to stop the system from sending metrics to the server when the user prefers not to disclose information by giving a flag to the installer.
StopMetrics()
```

patrickdillon replied:
I think we should limit this enhancement to the user interaction with the installer and the installer's interaction with Prometheus, i.e. the metrics design. @rna-afk will work on a separate enhancement solely focused on installer implementation/instrumentation. 

Regarding disabling metrics collection, in this doc we should discuss: is it a command-line flag, environment variable, install config setting, other? 

@abhinavdahiya let me know if that sounds acceptable to split this

brancz commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+
+## Open Questions
+
+ > 1. This requires creating an [aggregation pushgateway][weaveworks-aggregation-pushgateway] where the installer would push metrics. How would this be handled and by which team?
```
presumably the app-sre team, you're going to want to talk to them quickly though because their epic planning either just happened or is about to happen

brancz replied:
this might make sense to live in the telemeter-server which authenticates all other telemetry push requests already today

brancz replied:
I think this needs a rollout plan. However I would suggest that:

1) we add the instrumentation and just play around with the instrumentation and a local push gateway to understand it's use and intricacies (no production systems needed until we figure this out)
2) once validated everything works as we imagine we deploy the push aggregation gateway as part of openshift telemetry, but only configure CI to actually push to it to get some experience operating it
3) once we are comfortable running this as a production system we default to sending metrics to our production instance on every install

patrickdillon replied:
> presumably the app-sre team

I reached out to @pbergene 

> I think this needs a rollout plan. However I would suggest that:

That rollout plan sounds good. We can try to incorporate.

simonpasquier commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,300 @@
+cluster_installation_invocation{command="create", target="cluster",os="linux",platform="AWS",result="APIFailure",return="1",version="4.2",duration="30"} 1
+cluster_installation_invocation{command="destroy", target="cluster",os="linux",platform="AWS",result="Success",return="1",version="4.2",duration="10"} 1
+```
+* `duration` of the command represnts the overall time required to complete the command. The label values will be bounded in the installer client to preset buckets of a certain step-size, 2.5 in the example above.
````
s/represnts/represents/

simonpasquier replied:
Rather than encoding the duration bucket in the installer client, it might be more flexible/maintainable to perform the bucketing at the aggregation gateway level. Otherwise it will be impossible to adjust the buckets after the fact.

rna-afk replied:
```suggestion
* `duration` of the command represents the overall time required to complete the command. The label values will be bounded in the installer client to preset buckets of a certain step-size, 2.5 in the example above.
```

bwplotka replied:
And again, I think wit should be histogram as @brancz [mentioned](https://github.com/openshift/enhancements/pull/205/files#r380674219)

brancz commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+## Open Questions
+
+ > 1. This requires creating an [aggregation pushgateway][weaveworks-aggregation-pushgateway] where the installer would push metrics. How would this be handled and by which team?
+ > 2. If #1 is possible, how would client authentication be administered to prevent [malicious pushing of metrics](risks-and-mitigations)? Authentication based on what? A pull secret? 
```
1) yes I would like to see some basic forms of preventing malicious attacks (rate limits are probably sufficient)
2) pull secret auth is what we do today for openshift telemetry

patrickdillon replied:
Pull secret auth sounds reasonable but let's note a couple of minor limitations:

1. We won't be able to capture failed installs where the pull-secret is the problem.
2. Commands like gather and destroy don't necessarily depend on the pull secret
I think we could live/work with these limitations.

brancz replied:
makes sense, I doubt we want to accept any unauthenticated telemetry though

rna-afk replied:
@brancz do you know of any example code that authenticates with pull secret to push to Prometheus?

brancz commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+
+ > 1. This requires creating an [aggregation pushgateway][weaveworks-aggregation-pushgateway] where the installer would push metrics. How would this be handled and by which team?
+ > 2. If #1 is possible, how would client authentication be administered to prevent [malicious pushing of metrics](risks-and-mitigations)? Authentication based on what? A pull secret? 
+ > 3. How to filter to collect only real user usage and prevent OpenShift developer/Red Hat internal users from skewing metrics? Perhaps filter by version?
```
In any case we need to heavily confine the length of versions to produce very low cardinality something like `4.1.x`, `4.2.x`, `ci` etc. as all of this is going to be pushed to the push aggregation gateway as proposed which in turn will need to expose these on a metrics endpoint so we're talking an absolute upper bound within the 10 000s of time-series that we will ever be able to create for this, which we can reach very quickly.

brancz commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+ > 2. If #1 is possible, how would client authentication be administered to prevent [malicious pushing of metrics](risks-and-mitigations)? Authentication based on what? A pull secret? 
+ > 3. How to filter to collect only real user usage and prevent OpenShift developer/Red Hat internal users from skewing metrics? Perhaps filter by version?
+ > 4. Is it possible to determine whether a user modified an ignition config or manifest asset by comparing the asset the installer would generate to the one found on disk?
+ > 5. Are [package-scoped variables](internal-installer-design) acceptable in the proposed installer `metrics` package? 
```
Could you add more detail here? It's hard for me to tell whether this has anything to do with the actual metrics we send (I care about this) or internals of the code (I don't particularly care about :) ).

brancz commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+Some examples:
+
+
+As a telemetry analyst I would like to create queries which help determine how long the installation process takes in total and broken down by different stages of installation so that I can better understand user experience and expose any abnormalities.
```
We need to set appropriate expectations here from the start I think. The result will not be a full log and full understanding of a single run, but rather statistics about all runs. So we won't be able to answer "how long did this particular run take for this part", more 99% of all cluster installs finish within 40min.

derekwaynecarr replied:
how does this distinguish IPI flows versus UPI flows?  are we just instrumenting the create cluster command and not the wait for command?

patrickdillon replied:
> how does this distinguish IPI flows versus UPI flows? are we just instrumenting the create cluster command and not the wait for command?

The design is to implement `create`, `wait-for`, `gather`, and `destroy`. So we should capture UPI installs in wait-for, but have not yet made a distinction beyond that.

brancz commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,300 @@
+cluster_installation_invocation{command="create", target="cluster",os="linux",platform="AWS",result="APIFailure",return="1",version="4.2",duration="30"} 1
+cluster_installation_invocation{command="destroy", target="cluster",os="linux",platform="AWS",result="Success",return="1",version="4.2",duration="10"} 1
+```
+* `duration` of the command represnts the overall time required to complete the command. The label values will be bounded in the installer client to preset buckets of a certain step-size, 2.5 in the example above.
````
this sounds like a histogram, so we should follow the histogram rules for this, meaning we need to send the full histogram, with it's `le` buckets, otherwise we won't be able to do the queries that we intend to do

patrickdillon replied:
Initially I was put off because histogram bounds are less than or equal to. So if you have a 5 minute install, it would also go into the 10, 15, 20., and so on buckets... but you're right, we should change this.

brancz commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+
+Either of these proposed metrics would be pushed only upon successful completion of the `create cluster` command.
+
+#### Metric Labels
```
For every metric/label combinations that are proposed here, I would like to see a back of the envelope cardinality calculation. Even reading through this enhancement I have only a vague idea of the numbers. Especially for the back end side of things, capacity planning etc. this is very important.

brancz commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+
+By using the aggregation pushgateway, we are losing specificity. We will not be able to tell whether one user heavily skews results.
+
+ We want to understand user events in how they interact with the installer and get specific error messages and logs, but we cannot do that with Prometheus as a backing service. 
```
It's still a pretty long way to go I think, but there are plans to support log ingestion in the Observatorium system (the system that implements the metrics system today)

simonpasquier commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,300 @@
+# TYPE cluster_installation_modification counter
+cluster_installation_modification{asset="cluster-config.yaml,cluster-scheduler-02-config.yml"} 1
+```
+In this case, the values would be sorted in order to reduce cardinality. The benefit of this approach is that it creates a correlation between the modified resources in a particular run. It is less natural than the following approach and could make querying more difficult (feedback on querying, please).
````
If I read correctly, 30 files can be edited by the end-users. If values are concatenated into a single `asset` label, we can have many more values than 30 which would blow up cardinality.

rna-afk commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,300 @@
+```
+# HELP cluster_installation_modification lists all the assets that were changed from the original value by the user before the create command was called.
+# TYPE cluster_installation_modification counter
+cluster_installation_modification{asset="cluster-config.yaml,cluster-scheduler-02-config.yml"} 1
````
```suggestion
cluster_installation_modification{asset="cluster-config.yaml"} 1 ,
cluster_installation_modification{asset="cluster-scheduler-02-config.yml"} 1
```

bwplotka commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,300 @@
+### Implementation Details/Notes/Constraints
+The installer does not fit a typical Prometheus use case in that:
+* metrics will be pushed rather than scraped
+* meaning will come from metric labels as much as values
```
What does this mean? :thinking:

bwplotka commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,300 @@
+```
+* `duration` of the command represnts the overall time required to complete the command. The label values will be bounded in the installer client to preset buckets of a certain step-size, 2.5 in the example above.
+* `return` - denotes success or failure. It could potentially be subsumed in result.
+* `result` label denotes a command outcome:
````
Don't we want `type` to separate `create, waitfor`? commands?

patrickdillon commented:
@openshift/openshift-team-monitoring @abhinavdahiya @smarterclayton @crawford 

Thank you for your first round of reviews. I have taken the feedback into account and heavily revised the enhancement:

- Removed all discussion that is not directly related to the design of the metrics themselves. We will create separate documents discussing installer instrumentation and the implementation of the aggregation pushgateway
- Converted some of the metrics to histograms
- Provided cardinality calculations
- In the previous version, there was one metric capturing all installer commands. To reduce cardinality, this was broken into multiple metrics
- Created categories for modified assets

Many members of the installer team (including myself) have not yet had much exposure to Prometheus, so I tried to put together an FAQ to provide some context and serve as a sort of quickstart: https://github.com/patrickdillon/metrics-faq

patrickdillon commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+
+#### Commands
+
+##### cluster_installation_create 
```
`cluster_installation_create` does not seem like a great name to me. Any suggestions? 

We could perhaps follow a convention:
`cluster_installation_create_cluster`
`cluster_installation_create_manifests`
`cluster_installation_create_ignition`

lilic replied:
This is an info metric, IIUC? If yes, then going by the metric naming guidilenaces should be _info suffix e.g. :

> foobar_build_info (for a pseudo-metric that provides metadata about the running binary)

https://prometheus.io/docs/practices/naming/#metric-names

wking replied:
The metric is measuring invocation duration, right?  So from "should use base units..." and "should have a suffix describing the unit, in plural form..." sections of those metric-names docs, this should be `installer_create_seconds`, or something, with a `target="cluster"` (or manifests, or ignition-configs, etc.) label.

wking replied:
> The metric is measuring invocation duration, right?

Oops, I see that the value is just success (1) or failure (0), so `installer_create_info` or some such makes sense to me.

patrickdillon commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+* `target` and `result` are dependent:
+  - `bootstrap-complete`
+      - `BootstrapFailed`
+      - `APIFailed`
```
Is there a prometheus standard for casing or word separation?

lilic replied:
Yes these are define here https://prometheus.io/docs/practices/naming/#labels and https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

wking replied:
> Yes these are define here https://prometheus.io/docs/practices/naming/#labels ...

This specific line is for a value though, not the label name itself.  I don't see anything in the docs about casing conventions for values, and I'd recommend we use strings cased like we want to see in a plot legend.  Also, non-zero `wait-for` exits are just timeouts, not necessarily failures.  So maybe "bootstrap timeout", "Kubernetes API timeout", etc.?

patrickdillon commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+The proposed metrics represent categories of one or more asset file modified during an installation:
+
+Metric | Modified Files
+-------|---------------
```
@abhinavdahiya  could you give some feedback on what you think about these categories?

romfreiman replied:
why not to have this files as a tag instead? It would be much easier than maintain those categories over time and the number of files anyhow limited - so cardinatlity wise u either get different metric name or different tag. 
U can just submit metric per file as tag and that's it. Much easier to maintain

patrickdillon replied:
I don't understand what a "tag" is in this context. Can you elaborate?

romfreiman replied:
sorry, label

patrickdillon commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+      - `OperatorsDegraded`
+      - `Success`
+
+##### cluster_installation_manifests and cluster_installation_ignition
```
@abhinavdahiya is unclear to me whether we want to collect metrics for every invocation of `create install-config`, but if so I can add a simlar metric to this

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+## Open Questions
+
+ > 1. How to filter to collect only real user usage and prevent OpenShift developer/Red Hat internal users from skewing metrics? Perhaps filter by version?
+ > 2. Is it possible to determine whether a user modified an ignition config or manifest asset by comparing the asset the installer would generate to the one found on disk?
```
nit: "ignition" -> "Ignition".

And yeah, you can see which assets are dirty [here][1], although we don't currently cache that information on the `Asset` itself for future reference.

[1]: https://github.com/openshift/installer/blob/b518ee89d4ee64f95b3b9401b25439ce65480189/pkg/asset/store/store.go#L310-L315

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+    - `OperatorsDegraded`
+    - `Success`
+* `platform` - currently we deploy to 8 platforms: `aws`, `azure`, `gcp,` `libvirt`, `none`, `openstack`, `ovirt`, and `vsphere`. The list will grow. 
+* `os` - You can find all the available combinations of [Go OS and arch values](https://github.com/golang/go/blob/master/src/go/build/syslist.go) with `go tool dist list`. At this stage I assume we only want to colllect operating systems and some of the values are unlikely/impossible:
```
nit: "colllect" -> "collect"

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+  - `darwin` - this could be rewritten to `macos` on the installer client if we prefer
+  - `bsd` - this would encompass the `freebsd`, `netbsd`, and `openbsd` values
+  - `windows` - future proofing? 
+  - `other` - everything else not captured above
```
Why not just use GOOS directly?  I don't see a need to invent our own, separate naming system.

patrickdillon replied:
> Why not just use GOOS directly? 

The reasoning is to enforce a reduced cardinality. As we would practically only hit `linux` or `darwin` this reasoning may be flawed/over optimization. On the other hand this might be simple defensive programming.

wking replied:
> The reasoning is to enforce a reduced cardinality.

Cardinality is coming up in a number of these threads, but I'm not clear on the constraint.  There is the cardinality caution [here][1], but... folks can't be running the installer _that_ often, right?  And we already have running clusters reporting metrics with much greater cardinality.  Can we accept a higher cardinality for these installer metrics and then have a window (a week?  Two?) after which we erase old data?  It's not clear to me what sort of load an increased cardinality puts on our metrics pipelines; possibly different loads for any processing that happens at Telemetry-upload time vs. query-processing time.  It's also not clear to me how much capacity our current metrics pipelines have.  You calculate cardinality in a number of places in this proposal; is there a cap we need to stay under?  Do two metrics with cardinality A have the same load as a single metric with cardinality 2A?  Feel free to say "trust me, it's really serious to keep our cardinality low, and collapsing GOOS into fewer buckets is the only only way" without trying to talk me through it ;).

[1]: https://prometheus.io/docs/practices/naming/#labels

patrickdillon replied:
I appreciate your raising these points, as I have similar questions. I know that cardinality is a valid concern, but I don't know the actual capacity. Guidelines that I see leave it at cardinality should be bound ( then I look in telemter and see unbound cardinality :P).  Luckily the @openshift/openshift-team-monitoring has been giving this PR some attention and I think they can help point in the right direction. 

I do expect the installer will be considered a special case, but I think we are also abusing label so I am trying to keep this in mind.

The way I think about cardinality (I'm saying this to expose my logic/get confirmation at the risk of explaining the obvious) is that each time a metric is pushed with a unique combination of label key-value pairs, a time series is created. So we run the risk of having one-off time series created. Let's say there's a single user of I dunno `freebsd` that pushes an install metric, we will have an unchanging timeseries with a value of 1 monitored/scraped by prometheus indefinitely.

Assuming the formula I used is correct, my initial metrics proposal led to cardinality calculations over 200K. That just felt high so I was able to reorganize the metrics without losing much or any of the original data that would have been collected.

brancz replied:
@patrickdillon I think you understood the problem pretty well.

> then I look in telemter and see unbound cardinality

Where do you see this? The biggest point about all of this is our ability to somewhat plan for the expected load that these things would put on the system, as cardinality essentially directly translates into resource usage on the system (both for ingestion and querying). So even if there are theoretical unbound labels, they are bound in practice or in a way that we have planned into the capacity of the system.

patrickdillon replied:
> Where do you see this? 

I didn't mean to imply there was a problem or worse. I was thinking of something like cluster ID. Your phrase "bound in practice" describes this situation well.

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+  - `bsd` - this would encompass the `freebsd`, `netbsd`, and `openbsd` values
+  - `windows` - future proofing? 
+  - `other` - everything else not captured above
+* `version` - represents minor releases; assuming we do not backport metrics, `4.5` will be the only value if we make it into the next release and we could expect roughly 20 values over the next five years.
```
Existing metrics like `cluster_version` include the full version string in their `version` label (e.g. [here][1]).  Why would we not do the same here?

Also, where are we getting this version information from?  Currently `cluster_version` is using the version information baked into the `release-metadata` file in the release image's layers.  But also currently, the release image is an opaque pullspec as far as the installer is concerned.  Should this string be the installer version (leaning on openshift/installer#2682 and assuming no release image overrides)?  Or the pullspec's tag/digest?  Or...?

[1]: https://github.com/openshift/cluster-version-operator/blob/732ae625e991bf1447370b537caf9bcc22c850dc/pkg/cvo/metrics.go#L170

patrickdillon replied:
>  Why would we not do the same here?

In order to reduce cardinality, see e.g. https://github.com/openshift/enhancements/pull/205/files#diff-3b226c8b21f37f235bcd51fbc7128a3cR268

 > Also, where are we getting this version information from? 
> Should this string be the installer version (leaning on openshift/installer#2682 and assuming no release image overrides)? 

Great question. `version` should definitely be installer version. Will look into  openshilft/installer/#2682 as  a source. thanks for the suggestion.

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+  - `windows` - future proofing? 
+  - `other` - everything else not captured above
+* `version` - represents minor releases; assuming we do not backport metrics, `4.5` will be the only value if we make it into the next release and we could expect roughly 20 values over the next five years.
+* `le` - the standard Prometheus label for histogram buckets, this represents how long a command took complete. The current proposal is to have a 5 minute step starting at 15 minutes and maxing at 60.
```
It is not clear to me how this works if we eventually decide to adjust the buckets.  Is there a reason we can't use a [guage][1] and use Prometheus ranges to aggregate and bucket entries from different installer invocations?  Otherwise, if you want to add a new `le="65"` bucket, it seems like it would be hard to initialize the counter you're feeding, because you don't know how many of the previous runs which were > 60 but less than `+Inf` should go into the `le="65"` slot.

[1]: https://prometheus.io/docs/concepts/metric_types/#gauge

patrickdillon replied:
> It is not clear to me how this works if we eventually decide to adjust the buckets.

Once we set the step and range, we won't be able to adjust past data, so it's important we agree on the minimum level of granularity.   

>  Is there a reason we can't use a guage 

I'm not sure I understand this suggestion. The value of the metric will be the aggregate count of installer metrics that fall into the duration bucket specified in the label--otherwise, using the installer duration as a value, doesn't make sense as a time series or would otherwise be difficult to get a count from.   I tried to explain the difficulty in using duration as a metric value here: https://github.com/patrickdillon/metrics-faq#standard-pushgateway-basic-example If the suggestion is to still use buckets, then that is essentially a histogram.

wking replied:
> Once we set the step and range, we won't be able to adjust past data, so it's important we agree on the minimum level of granularity.

It's not just the minimum, right?  We can't change our mind about buckets ever.  Or is there a change that we can make that won't break compat with the data that Prometheus has already slurped in?

> I tried to explain the difficulty in using duration as a metric value here: https://github.com/patrickdillon/metrics-faq#standard-pushgateway-basic-example ...

I had not seen this before; useful stuff.  I hadn't realized the Prometheus histograms were a client-side construct, with buckets being chosen when the metric was defined (e.g. [here][1]).  One way to address your "Each change in the graph denotes a sample pushed to the gateway. But what does this graph tell us?" might be to push a guage with "I took 17 minutes", etc., and then use [aggregation][2] with something like:

```
quantile_over_time(0.95, cluster_installation_create_seconds[7d])
```

to get the 95% quantile for installation time over the past week.  Or with a specific threshold:

```
count_over_time(cluster_installation_create_seconds[7d] > 30*60) / 
count_over_time(cluster_installation_create_seconds[7d])
```

to get the fraction of installs over 30m in the past week.  Or does that not work for some reason?

[1]: https://godoc.org/github.com/prometheus/client_golang/prometheus#HistogramOpts
[2]: https://prometheus.io/docs/prometheus/latest/querying/functions/#aggregation_over_time

brancz replied:
`quantile_over_time` or `count_over_time` would not be sufficient as they act on single time-series `over_time`, whereas the histogram is a representation of all runs over time. Remember that the push aggregation gateway would still be scraped by Prometheus only every x seconds, so we don't have every push event in Prometheus but only the snapshot of the counter when Prometheus sees it at scrape time.

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+* `le` - the standard Prometheus label for histogram buckets, this represents how long a command took complete. The current proposal is to have a 5 minute step starting at 15 minutes and maxing at 60.
+
+##### cluster_installation_waitfor
+UPI installs utilize the wait-for command, which would be tracked with a similar histogram:
```
Is this only a separate metric because of the different, hard-coded bucket boundaries?  Otherwise it seems easier to have a single `installer_command_info` with the command (`create cluster`, `wait-for bootstrap-complete`) as a `command` label or some such.

patrickdillon replied:
> Is this only a separate metric because of the different, hard-coded bucket boundaries?

Yes, the reduced bucket labels reduces the cardinality by 800 time series. (more if you account for versions).

wking commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,430 @@
+cluster_installation_ignition{platform="aws",result="success",version="4.2",os="linux"} 1
+```
+
+Possible `result` values are `success` and are `error`, otherwise all possible label values are the same as `cluster_installation_create`. The sample value would always be one in order to aggregate to a total count of runs of this command.
````
Some of your metrics use `success` and some use `Success`.  You should probably pick one way for all the metrics.

wking replied:
Also, `success` and `error` are hard to extend later.  For example, you could sum `result="success"` over the whole to get a success rate.  But if you sum `result="error"` over the whole to get an error rate, that would break down if we subsequently split `error` into sub-cases (e.g. `throttled checking credentials`).  I think it's safer to have an `error` label that we set to the empty string for success, and which we set to `other` for uncategorized errors.

patrickdillon replied:
I agree with standardizing capitalization. Wouldn't you just query for `not success`?

wking replied:
> Wouldn't you just query for `not success`?

That would work.  I just think `result="error"` is a foot-gun, and would rather not have it around ;).  `result="other error"` avoids the foot-gun, but feels (subjectively) awkward.  `error="none"` and `error="other"` feel good to me, but maybe they feel awkward to you?

wking commented on enhancements/installer/telemeter-metrics.md:
````diff
@@ -0,0 +1,430 @@
+cluster_installation_destroy_sum{platform="aws",result="success",version="4.2",os="linux"} 10
+cluster_installation_destroy_count{platform="aws",result="success",version="4.2",os="linux"} 1
+```
+The label values are the same as `openshift_installation_create` except `le` has values between 5 and 30 minutes and `result` is `success` or `error`.
````
Do we want to distinguish between "installer decided to error out" (e.g. we hit an error that we know we can't recover from) and "received a SIGTERM" (e.g. user got tired of waiting and gently shut us down)?

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+* `cluster_installation_duration_operators`
+
+Each of these metrics (except infrastructure) would have five labels:
+  * `command` - either `create` or `wait-for`
```
The semantics of this for `wait-for` are not clear to me.  Is `cluster_installation_duration_api{command="wait-for"}` supposed to be "since calling [`waitForBootstrapComplete`][1]" or "since something we expect to have been the previous step in a user-provided-infrastructure flow, e.g. `create ignition-configs`"?

[1]: https://github.com/openshift/installer/blob/b518ee89d4ee64f95b3b9401b25439ce65480189/cmd/openshift-install/waitfor.go#L47

patrickdillon replied:
It is supposed to be since `openshift install wait-for <target>` was called.

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+cluster_installation_modification_machineconfig_manifest | 04-openshift-machine-config-operator.yaml <br> machine-config-server-tls-secret.yaml
+cluster_installation_modification_ca_manifest | machine-config-server-tls-secret.yaml
+cluster_installation_modification_pullsecret_manifest | openshift-config-secret-pull-secret.yaml
+cluster_installation_modification_bootstrap_ignition | bootstrap.ign
```
I doubt the installer will reliably have access to the actual bootstrap Ignition config, because it's so easy for a user to pick our generated content up after `create ignition-configs` and then modify it elsewhere on the way to publishing where the bootstrap machine will be able to find it.  Do we want to publish the metric, even though that flow would show up (incorrectly) as a "`bootstrap.ign` was not modified"?  And who would be checking to see if `bootstrap.ign` was modified anyway?  Every `wait-for` command getting invoked?  I'd rather just drop these and have metrics only for files/assets sucked in by the installer in the asset graph leading to the Ignition config files.

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+cluster_installation_modification_network_manifest | cluster-network-01-crd.yml <br> cluster-network-02-config.yml <br> cluster-proxy-01-config.yaml
+cluster_installation_modification_scheduler_manifest | cluster-scheduler-02-config.yml
+cluster_installation_modification_cvo_manifest | cvo-overrides.yaml
+cluster_installation_modification_etcd_manifest | etcd-ca-bundle-configmap.yaml <br> etcd-client-secret.yaml <br> etcd-host-service-endpoints.yaml <br> etcd-host-service.yaml <br> etcd-metric-client-secret.yaml <br> etcd-metric-serving-ca-configmap.yaml <br> etcd-metric-signer-secret.yaml <br> etcd-namespace.yaml <br> etcd-service.yaml <br> etcd-serving-ca-configmap.yaml <br> etcd-signer-secret.y ...
```
Do we not care about installer version for these metrics?

Also, can we not have a single metric `installer_asset_modified` with an `asset` label and value from `Asset.Name()`?  That way we could push this metric when we load each asset (and determine [whether is has been modified or not][1], instead of growing the more granular ability to track dirty files (vs. assets) on load and then iterating over dirty files to figure out which non-asset file groupings were modified or not.

[1]: https://github.com/openshift/enhancements/pull/205#discussion_r391313129

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+
+The pushgateway would be an a public endpoint reachable on the internet. Therefore some authentication or security should be in place in order to prevent malicious pushing of metrics.
+
+Users and customers may be wary of exposing information when using the installer.
```
is there an opt-out mitigation plan for this?

tsorya commented:
@patrickdillon have a small question. 
Can this flow be used to report installation progress?

patrickdillon commented:
> Can this flow be used to report installation progress?

@tsorya No, not with this proposal. The current design is that the metrics would be pushed in a batch at the end of an installation. Moreover, if you are interested in the the progress of a particular installation, this enhancement would not help because all metrics are designed to be **aggregated** so there are not any unique identifiers.

On the other hand, metrics reported by the cluster version operator might be able to help you with installations in progress once the installation has progressed past infrastructure provisioning and the CVO is up.

tsorya commented:
@patrickdillon thanks for explanation

wking commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+
+By using the aggregation pushgateway, we are losing specificity. We will not be able to tell whether one user heavily skews results.
+
+ We want to understand user events in how they interact with the installer and get specific error messages and logs, but we cannot do that with Prometheus as a backing service. 
```
nit: leading whitespace.

crawford commented:
I think the set of metrics described here make sense, but can we include the list of questions that we are trying to answer with this telemetry? For example, "how many of our customers are using Linux vs MacOS?" or "what are the most commonly used targets and which ones are typically used in conjunction?". We'll want to make sure that we can actually write the relevant PromQL queries given the metrics we are collecting.

patrickdillon commented:
> I think the set of metrics described here make sense, but can we include the list of questions that we are trying to answer with this telemetry? For example, "how many of our customers are using Linux vs MacOS?" or "what are the most commonly used targets and which ones are typically used in conjunction?". We'll want to make sure that we can actually write the relevant PromQL queries given the metrics we are collecting.

Absolutely. I have some old notes to this effect, but looking at them now they are outdated. We can come up with some better queries now that the metrics have evolved through feedback.

staebler commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+* `le` - the standard Prometheus label for histogram buckets, this represents how long a command took complete. The current proposal is to have a 5 minute step starting at 15 minutes and maxing at 60.
+
+##### cluster_installation_waitfor
+UPI installs utilize the wait-for command, which would be tracked with a similar histogram:
```
When an install fails, Hive uses `wait-for` commands to (1) check whether bootstrapping succeeded and, in some cases, (2) wait longer for cluster initialization to complete. Maybe we could add a flag for the `wait-for` commands to not do metrics gathering so that Hive's use cases do not skew results.

staebler commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+
+##### cluster_installation_destroy
+
+When a user destroys a cluster a single sample of this metric would be pushed:
```
Hive may run the cluster destroy twice when a cluster install fails. I do not know if we care about any skew in results in that case.

dgoodwin commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+
+As a telemetry analyst I would like to create queries which analyze aggregate installation data to determine how long the installation process takes in total and broken down by different stages so that I can better understand user experience, expose any abnormalities, and compare different versions of the installer.
+
+As a telemetry analyst I would like to create queries which show aggregate counts of failure by stage of installation so that we can better understand where users are experiencing problems. (This will not provide  specific error messages/logs).
```
In https://github.com/openshift/installer/pull/3535/files the condition "reason" keys would be extremely useful to report and shouldn't be a problem for cardinality. Being able to see the rate at which specific known problems are occurring would be immensely valuable.

dgoodwin commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+
+As a telemetry analyst I would like to see how often users modify ignition configs or manifests so that I can better understand how users are interacting with the installer. [(see open question #3)](open-questions)
+
+As a telemetry analyst I would like to see how often the installer is being run on certain operating systems so I can better understand our user environments and needs.
```
Per our convo over in Jira, would love to see OPENSHIFT_INSTALL_INVOKER reported so we can see hive provisions in the wild.

patrickdillon replied:
@dgoodwin the value for INVOKER is `user` if unset or the value from the env var. How can we bound the cardinality but still provide value in the metrics? Would it be enough to do a binary classification to say "these are all the installs where INVOKER is set (not `user`)"? 

If we need more detail than that, can you provide some input on how the `INVOKER` value is formatted?

dgoodwin replied:
This doesn't strike me as a use case where cardinality would be an issue. We should only be dealing with a handful of values, user, hive, managed azure... maybe a dozen? It's not something that would be driven by user input or random values in any normal operation. Is there a concern over some malicious or overly enthusiastic user finding the var and setting it to random values?

Just to clarify are we expecting a metric like openshift_install_invoker{type=user} = 1? 

Could we trim this down to type=unknown if we don't recognize the type in the aggregation.

patrickdillon replied:
>  maybe a dozen? It's not something that would be driven by user input or random values in any normal operation.

sounds good. I am (or was, until now) ignorant of the values. Are these values defined in hive code somewhere? I think the best way it would work is if hive & the installer had some sort of shared enumeration of potential values, similar in concept to how values (for e.g. operating system) are enumerated here: https://github.com/openshift/enhancements/blob/2196381b04834a96b15a739611be14d46a200d9b/enhancements/installer/telemeter-metrics.md#label-values

> . Is there a concern over some malicious or overly enthusiastic user finding the var and setting it to random values?

It seems like an unlikely threat but is so easy to guard against we should definitely do it to avoid any repercussions to production systems. If it is a nuisance to have a shared enumeration between installer & hive we could rethink it.

> Just to clarify are we expecting a metric like openshift_install_invoker{type=user} = 1?

No. I was thinking more along the lines of we would add it to any or all of the metrics suggested here that would be useful. If we added invoker to the cluster creation metric we could have samples like this for for the 15 minute or less installation bucket:
```
cluster_installation_create_bucket{invoker="user",platform="azure",result="success",version="4.6",os="linux",le="15"} 0
cluster_installation_create_bucket{invoker="managed-azure",platform="azure",result="success",version="4.6",os="linux",le="15"} 3
```
This would allow us to use all these metrics and select along an invoker dimension.

wking replied:
> This would allow us to use all these metrics and select along an invoker dimension.

No need for this.  A single metric with the invoker label can be joined with other metrics at query-time using `_id` (the cluster's Telemetry ID).

dgoodwin replied:
I would expect Hive would just have one hardcoded value we submit as invoker for any cluster install started via Hive. I'd really just be looking for trends, how many installs are being done through Hive's over time and their success rates.

romfreiman replied:
we also using the OPENSHIFT_INSTALL_INVOKER='assisted-installer' for clusters that will be created in such manner

patrickdillon commented:
Is that possible? I don't think we will be including cluster telemeter IDs
in these metrics (would love to be wrong about this).

On Thu, Jun 18, 2020, 10:33 PM W. Trevor King <notifications@github.com>
wrote:

> *@wking* commented on this pull request.
> ------------------------------
>
> In enhancements/installer/telemeter-metrics.md
> <https://github.com/openshift/enhancements/pull/205#discussion_r442599165>
> :
>
> > +3. Pushgateway implementation will be discussed in a separate design/enhancement
> +
> +## Proposal
> +
> +### User Stories
> +
> +#### Telemetry Analyst
> +Some examples:
> +
> +As a telemetry analyst I would like to create queries which analyze aggregate installation data to determine how long the installation process takes in total and broken down by different stages so that I can better understand user experience, expose any abnormalities, and compare different versions of the installer.
> +
> +As a telemetry analyst I would like to create queries which show aggregate counts of failure by stage of installation so that we can better understand where users are experiencing problems. (This will not provide  specific error messages/logs).
> +
> +As a telemetry analyst I would like to see how often users modify ignition configs or manifests so that I can better understand how users are interacting with the installer. [(see open question #3)](open-questions)
> +
> +As a telemetry analyst I would like to see how often the installer is being run on certain operating systems so I can better understand our user environments and needs.
>
> This would allow us to use all these metrics and select along an invoker
> dimension.
>
> No need for this. A single metric with the invoker label can be joined
> with other metrics at query-time using _id (the cluster's Telemetry ID).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/enhancements/pull/205#discussion_r442599165>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAVQYPAWY4GSBZAMLXZY5DRXLE6VANCNFSM4KRU5TEA>
> .
>

wking commented:
> Is that possible? I don't think we will be including cluster telemeter IDs
in these metrics (would love to be wrong about this).

Ah, maybe we generate that too late.  Is there another ID we could use to string these together?

patrickdillon commented:
IDs on the metrics would create a unique time series for each run of the
installer. That would result in unbound cardinality and be counter to the
aggregation approach proposed here.

Here is a background thread where I discussed with Frederic and Lili:
https://coreos.slack.com/archives/C0VMT03S5/p1575988237117400?thread_ts=1575988237.117400&cid=C0VMT03S5

On Thu, Jun 18, 2020, 11:32 PM W. Trevor King <notifications@github.com>
wrote:

> Is that possible? I don't think we will be including cluster telemeter IDs
> in these metrics (would love to be wrong about this).
>
> Ah, maybe we generate that too late. Is there another ID we could use to
> string these together?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/enhancements/pull/205#issuecomment-646415945>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAVQYLQH3T7ZRCIOYUA5STRXLL4DANCNFSM4KRU5TEA>
> .
>

romfreiman commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+    - `APIFailed`
+    - `OperatorsDegraded`
+    - `Success`
+* `platform` - currently we deploy to 8 platforms: `aws`, `azure`, `gcp,` `libvirt`, `none`, `openstack`, `ovirt`, and `vsphere`. The list will grow. 
```
baremetal?

romfreiman commented:
If we are able to generate cluster uuid as a first step in the installer, than we'll be able to remove the push gateway (and all the operational burden around it). Then we'll mark every execution of the installer with same uuid (it's fine that we cannot identify multiple executions - only by timestamp).  
Also, we'll be able to track a history of cluster - for example, user executes installer at least 5 times before a success
 - with current design - we''ll not have this data - it might be that there a user that fails 1000 times and all the rest have 100% success rate - but we'll never know.
So why not to generate the cluster id as the first step in the installer and remove all the complexity.
@brancz I know you have concerns about cardintality, but it will not be different from any other metric that is generated by OCP clusters. Also, amount wise, I assume number of metrics submitted by the installer are significantly lower than any other live system
cc: @patrickdillon

brancz commented:
@patrickdillon did an initial calculation with per cluster metrics, which turned out to be so huge that it was decided against (as in a multiple of all current data being sent if I recall correctly). Maybe we can include that in this enhancement in the alternatives section?

romfreiman commented:
@brancz how is it different from any other OCP metric? All of them has cluster_id as one of the labels

patrickdillon commented:
> @patrickdillon did an initial calculation with per cluster metrics, which turned out to be so huge that it was decided against (as in a multiple of all current data being sent if I recall correctly).

I have not done any cardinality calculations which would include cluster id, because cluster id is essentially unbound. Cluster id is generated along with the ignition-configs and would be unique for each attempted install.  There might be some minor exceptions to this (e.g. `create cluster` then `wait-for`), but I think it makes sense to treat it as unique. 

> Also, we'll be able to track a history of cluster - for example, user executes installer at least 5 times before a success

I don't think we would be able create a link between installs in this case. The typical workflow for repeating an install is to remove the dir and then create again, which would generate a new cluster id. In most cases a user is blocked from reusing the same cluster id when creating a cluster because the installer will detect `terraform.tfstate` and exit. (cc @abhinavdahiya  & @wking  for fact checking this and above point)

> how is it different from any other OCP metric? All of them has cluster_id as one of the labels

We discussed this a bit here: https://github.com/openshift/enhancements/pull/205#discussion_r393708220 This case is different because a new time series would be created for each attempted install, where in telemeter cluster ID is bound by active cluster (or, perhaps, historically active cluster).

> I assume number of metrics submitted by the installer are significantly lower than any other live system

I think the relevant concern for cardinality is # of time series created, rather than number of samples pushed. If we include cluster id, installer will only push a single sample, but each sample will create a new time series with an unchanging value, which will be perpetually scraped .

@brancz given that cluster id would be essentially unique, this is a non-starter due to cardinality concerns, right? if we could include cluster id, that would definitely simplify some aspects (we could do away with aggregation gateway), and we should consider it, but based on slack conversation linked above I think we need to keep aggregation in order to create a sensible time series.

rna-afk commented on enhancements/installer/telemeter-metrics.md:
```diff
@@ -0,0 +1,430 @@
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="15"} 0
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="20"} 0
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="25"} 0
+cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="30"} 1
```
```suggestion
cluster_installation_create_bucket{platform="aws",result="success",version="4.2",os="linux",le="30"} 0
```
example below says between 30-35 mins

patrickdillon replied:
Hm, isn't this largely correct?

Technically, I think this depends on the [rounding in the installer](https://golang.org/pkg/time/#Duration.Round). So if an install took 29 minutes and 30 seconds, the value would be rounded to 30 and `le="30"` (less than or equal to 30) would be true. So really the window should be 29:30-34:29 seconds, inclusive. 

Do correct me if I'm wrong.

rna-afk replied:
Ah my bad. Didn't see the sum part.
